### PR TITLE
Dpdk concurrent hash support

### PR DESCRIPTION
**Pull request by amarsri28**
This is the functionality for adding support for Dpdk hash library in bess for concurrent access. Unit test and sim mode test is passing.

### CODE REVIEW

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by omecproject** [comment]
Can one of the admins verify this patch?

**Comment by krsna1729** [comment]
test this please

**Comment by krsna1729** [comment]
retest this please

File: patches/0012-Dpdk-concurrent-hash-support-in-bess.patch [hunk 1]
```diff
@@ -0,0 +1,681 @@
...
++					if ((*hit_mask & (1ULL << k))  == 0) 
++           {
++					vals[k]= default_value;
++					}
++          }
++      
++     }
++     
+     }
+   }
```
Comment (krsna1729) [inline patches/0012-Dpdk-concurrent-hash-support-in-bess.patch]: better alignment

File: patches/0012-Dpdk-concurrent-hash-support-in-bess.patch [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+@@ -283,11 +331,29 @@ class ExactMatchTable {
+   void Find(const ExactMatchKey *keys, T *vals, size_t n,
+             T default_value) const {
+     const auto &table = table_;
+-    for (size_t i = 0; i < n; i++) {
+-      const auto *entry =
+-          table.Find(keys[i], ExactMatchKeyHash(total_key_size_),
+-                     ExactMatchKeyEq(total_key_size_));
+-      vals[i] = entry ? entry->second : default_value;
++    uint64_t *hit_mask;
```
Comment (krsna1729) [inline patches/0012-Dpdk-concurrent-hash-support-in-bess.patch]: uint32_t is enough?

**Comment by krsna1729** [comment]
in pfcpiface/upf.go remove codeblocks with pauseall and resumeall.

File: docker_setup.sh [hunk 3]
```diff
@@ -158,7 +158,7 @@ docker run --name bess -td --restart unless-stopped \
 	--net container:pause \
 	$PRIVS \
 	$DEVICES \
-	upf-epc-bess:"$(<VERSION)" -grpc-url=0.0.0.0:$bessd_port
+	upf-epc-bess:"$(<VERSION)" -m 4096 -grpc-url=0.0.0.0:$bessd_port
```
[the remark this answers is not in the document]
Comment (krsna1729) [reply docker_setup.sh:161]: can you remove this @amarsri28 ?

**Comment by krsna1729** [comment]
@amarsri28 is the sim mode test working correctly for you. With 0 rules inserted, packets should be dropped. Instead they are going out of gate 0

image

File: patches/0012-Dpdk-concurrent-hash-support-in-bess.patch [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+diff --git a/core/modules/wildcard_match.cc b/core/modules/wildcard_match.cc
+index ec639a7b..db501f12 100644
+--- a/core/modules/wildcard_match.cc
++++ b/core/modules/wildcard_match.cc
+@@ -56,6 +56,22 @@ static inline int is_valid_gate(gate_idx_t gate) {
+   return (gate < MAX_GATES || gate == DROP_GATE);
+ }
+ 
++
++void WildcardMatch::Initkeys(wm_hkey_t *keys)//,size_t key_size)
```
[the remark this answers is not in the document]
Comment (amarsri28) [reply patches/0012-Dpdk-concurrent-hash-support-in-bess.patch]: debug commented log removed

File: patches/0012-Dpdk-concurrent-hash-support-in-bess.patch [hunk 5]
```diff
@@ -0,0 +1,681 @@
...
+     mask(&key_masked, key, tuple.mask, total_key_size_);
+ 
+-    const auto *entry =
+-        ht.Find(key_masked, wm_hash(total_key_size_), wm_eq(total_key_size_));
+-
+-    if (entry && entry->second.priority >= result.priority) {
+-      result = entry->second;
++      WmData* entry =nullptr;
++      Initkeys(&key_masked); //u64_arr
++//      std::cout << "Wildcard:Find"<< "key="<<key_masked.u64_arr[0] << ","<< key_masked.u64_arr[1]<< ","<<key_masked.u64_arr[2] << ","<<key_masked.u64_arr[3] << ","<< key_masked.u64_arr[4] << ","<< key_masked.u64_arr[5] << ","<<key_masked.u64_arr[6] << ","<<key_masked.u64_arr[7] << std::endl;
```
[the remark this answers is not in the document]
Comment (amarsri28) [reply patches/0012-Dpdk-concurrent-hash-support-in-bess.patch]: debug logs were kept intentionally but now removed in new patch.
Its ok to have a single comment for repeated cosmetic issues.

File: patches/0012-Dpdk-concurrent-hash-support-in-bess.patch [hunk 6]
```diff
@@ -0,0 +1,681 @@
...
+--- a/core/utils/exact_match_table.h
++++ b/core/utils/exact_match_table.h
+@@ -155,7 +155,18 @@ typedef std::vector<std::vector<uint8_t>> ExactMatchRuleFields;
+ template <typename T>
+ class ExactMatchTable {
+  public:
+-  using EmTable =
++ struct rte_hash_parameters dpdk_params1 
++ {
++      .name= "test1",
```
[the remark this answers is not in the document]
Comment (amarsri28) [reply patches/0012-Dpdk-concurrent-hash-support-in-bess.patch]: done.

File: patches/0012-Dpdk-concurrent-hash-support-in-bess.patch [hunk 7]
```diff
@@ -0,0 +1,681 @@
...
+                              ExactMatchKeyEq(total_key_size_));
+@@ -271,10 +308,21 @@ class ExactMatchTable {
+   // Find an entry in the table.
+   // Returns the value if `key` matches a rule, otherwise `default_value`.
+   T Find(const ExactMatchKey &key, const T &default_value) const {
+-    const auto &table = table_;
+-    const auto *entry = table.Find(key, ExactMatchKeyHash(total_key_size_),
+-                                   ExactMatchKeyEq(total_key_size_));
+-    return entry ? entry->second : default_value;
++    //const auto &table = table_;
```
[the remark this answers is not in the document]
Comment (amarsri28) [reply patches/0012-Dpdk-concurrent-hash-support-in-bess.patch]: Its ok to have a single comment for repeated cosmetic issues.

File: patches/0012-Dpdk-concurrent-hash-support-in-bess.patch (hunk 2 above)
Comment (amarsri28) [reply patches/0012-Dpdk-concurrent-hash-support-in-bess.patch]: dpdk expect uint64_t

File: docker_setup.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (amarsri28) [reply docker_setup.sh:161]: done. added as we needed huge page, but seems like current config already hash that.

File: patches/0012-Dpdk-concurrent-hash-support-in-bess.patch [hunk 8]
```diff
@@ -0,0 +1,681 @@
...
++    if(i)
++    {
++      int j = i/sizeof(keys->u64_arr[0]);
++      int total_size =  sizeof(wm_hkey_t)/sizeof(keys->u64_arr[0]);
++      for(int i= total_size -j; i<total_size;i++)
++      {
++        keys->u64_arr[i] = 0;
++      }
++    }
++  }
```
[the remark this answers is not in the document]
Comment (amarsri28) [reply patches/0012-Dpdk-concurrent-hash-support-in-bess.patch]: memset would be faster . done

**Comment by krsna1729** [comment]
@amarsri28 also incoming and outgoing values are not equal at PDRLookup

**Comment by amarsri28** [comment]
> @amarsri28 also incoming and outgoing values are not equal at PDRLookup

fixed this issue
image

**Comment by amarsri28** [comment]
image


> > @amarsri28 also incoming and outgoing values are not equal at PDRLookup
> 
> fixed this issue
> image

**Comment by krsna1729** [comment]
retest this please

**Comment by yoooou** [comment]
retest this please

**Comment by yoooou** [comment]
retest this please

**Comment by yoooou** [comment]
retest this please

**Comment by krsna1729** [comment]
@yoooou what made it pass?

**Comment by krsna1729** [comment]
@amarsri28 rebase on latest master

**Pull request closed without merging**